### PR TITLE
Preserve requires_grad in ArticulationView attribute getters

### PR DESCRIPTION
**Pull request by jvonmuralt**
## Newton Migration Guide

Please ensure the migration guide for **warp.sim** users is up-to-date with the changes made in this PR.

- [ ] The migration guide in ``docs/migration.rst`` is up-to date

## Before your PR is "Ready for review"

- [ ] Necessary tests have been added and new examples are tested (see `newton/tests/test_examples.py`)
- [ ] Documentation is up-to-date
- [ ] Code passes formatting and linting checks with `pre-commit run -a`

## Summary by CodeRabbit

* **Bug Fixes**
  * Preserve autograd connectivity for selected/reshaped attributes (including indexed selections), ensuring gradients propagate correctly through materialized views and empty selections are handled safely.
* **New Features**
  * Add differentiable indexed-to-contiguous materialization so non-contiguous selections participate in forward/backward passes.
* **Tests**
  * Added tests verifying requires_grad propagation and backward-gradient updates for contiguous and indexed DOF/attribute selections.

### CODE REVIEW

File: newton/_src/utils/selection.py [hunk 1]
```diff
@@ -1188,7 +1207,7 @@ def _get_attribute_array(self, name: str, source: Model | State | Control, _slic
 
     def _get_attribute_values(self, name: str, source: Model | State | Control, _slice: slice | None = None):
         attrib = self._get_attribute_array(name, source, _slice=_slice)
-        if hasattr(attrib, "_staging_array"):
+        if hasattr(attrib, "_staging_array") and not getattr(attrib, "requires_grad", False):
```
Comment (nvlukasz) [inline newton/_src/utils/selection.py]: Not sure if this logic is correct. We want to always return the latest values even if the attribute has a gradient, right?

File: newton/tests/test_selection.py [hunk 2]
```diff
@@ -1318,5 +1318,66 @@ def test_multiple_articulations_per_world(self):
         assert_np_equal(stiffness.numpy(), expected)
 
 
+class TestArticulationViewRequiresGrad(unittest.TestCase):
```
Comment (eric-heiden) [inline newton/tests/test_selection.py:1327]: It would be good to have an actual test for gradients here where one Model/State array has a `array.grad` defined that is nonzero, and then the selected version of this array should have the same gradient.

File: newton/_src/utils/selection.py [hunk 3]
```diff
@@ -1189,10 +1247,23 @@ def _get_attribute_array(self, name: str, source: Model | State | Control, _slic
     def _get_attribute_values(self, name: str, source: Model | State | Control, _slice: slice | None = None):
         attrib = self._get_attribute_array(name, source, _slice=_slice)
         if hasattr(attrib, "_staging_array"):
-            wp.copy(attrib._staging_array, attrib)
```
Comment (nvlukasz) [inline newton/_src/utils/selection.py:1192]: Note that `wp.copy()` also copies gradients when both the source and destination arrays have gradients. Is it not sufficient to call `wp.copy(attrib._staging_array, attrib)` here? Do we need the custom code below?

Comment (jvonmuralt) [reply newton/_src/utils/selection.py:1192]: I do not think it is sufficient, what I understood from my experiments:
 
 1. indexedarray has no .grad, after attrib[slices] with non-contiguous indices, attrib becomes a wp.indexedarray, which does not carry a .grad attribute. So wp.copy() has no source gradient to copy from. Minimal script to demonstrate it:

_Create a source array with requires_grad=True_
src = wp.array([1.0, 2.0, 3.0, 4.0, 5.0], dtype=float, requires_grad=True)
print(f"src type:          {type(src).__name__}")
print(f"src.requires_grad: {src.requires_grad}")
print(f"src has .grad:     {hasattr(src, 'grad') and src.grad is not None}")

_Index it with non-contiguous indices → becomes wp.indexedarray_
indices = wp.array([0, 2, 4], dtype=int)
indexed = src[indices]
print(f"\nindexed type:          {type(indexed).__name__}")
print(f"indexed has .grad:     {hasattr(indexed, 'grad')}")
print(f"indexed.data type:     {type(indexed.data).__name__}")
print(f"indexed.data has .grad:{hasattr(indexed.data, 'grad') and indexed.data.grad is not None}")

_wp.copy into a staging array_
staging = wp.empty_like(indexed)
print(f"\nstaging.requires_grad before copy: {staging.requires_grad}")
wp.copy(staging, indexed)
print(f"staging.requires_grad after copy:  {staging.requires_grad}")
print(f"staging data: {staging.numpy()}")

_Even if we set requires_grad on staging, wp.copy can't copy grad from indexedarray_
staging2 = wp.zeros_like(indexed, requires_grad=True)
wp.copy(staging2, indexed)
print(f"\nstaging2.requires_grad: {staging2.requires_grad}")
print(f"staging2.grad:          {staging2.grad.numpy()}")
print(f"(grad is all zeros — wp.copy had no source .grad to copy from)")


2. wp.copy() is not tape-recorded, even if we could get the forward gradient copied, wp.copy is invisible to wp.Tape, so tape.backward() cannot propagate gradients back through it.
The custom gather kernel solves it, it takes the pre-indexed wp.array (which has .grad) as input, and because it's launched via wp.launch, the tape records it and auto-generates the correct adjoint for backward propagation. The separate wp.copy of .grad values is only for forward gradient visibility. 


There might be a better way to achieve this though, that I am not aware of.

Comment (preist-nvidia) [reply newton/_src/utils/selection.py:1192]: @jvonmuralt @nvlukasz what is the status here can we merge?

Comment (nvlukasz) [reply newton/_src/utils/selection.py:1192]: Let's merge
